File: Leetcode_with_Cpp/887superEggDrop.cpp
class Solution {
public:
    int superEggDrop(int K, int N) {
        vector<vector<int>> dp(N + 1, vector<int>(K + 1, 0));
        for (int i = 0; i <= N; i++) dp[i][1] = i;
        for (int j = 1; j <= K; j++) dp[1][j] = 1;
        for (int i = 1; i <= N; i++) {
            for (int j = 2; j <= K; j++) {
                int res = INT_MAX;
                /*
                for (int floor = 1; floor <= i; floor++)
                    res = min(res, 1 + max(dp[floor - 1][j - 1], dp[i - floor][j]));
                dp[i][j] = res;
                */
                int start = 1, end = i, mid;
                while (start <= end) {
                    mid = start + (end - start) / 2;
                    if (dp[mid - 1][j - 1] == dp [i - mid][j]) {
                        res = min(res, 1 + dp[mid - 1][j - 1]);
                        break;
                    } else if (dp[mid - 1][j - 1] > dp[i - mid][j]) {
                        end = mid - 1;
                        res = min(res, 1 + dp[mid - 1][j - 1]);
                    } else {
                        start = mid + 1;
                        res = min(res, 1 + dp[i - mid][j]);
                    }
                }
                dp[i][j] = res;
            }
        }
        return dp[N][K];
    }
};

File: py/887superEggDrop.py
class Solution(object):
    def superEggDrop(self, K, N):
        memo = {}
        def dp(K, N):
            if K == 1:
                return N
            if N == 0:
                return 0
            if N == 1:
                return 1
            if (K, N) in memo:
                return memo[(K, N)]
            res = float("inf")
            # 线性搜索穷举每一层楼扔鸡蛋的结果
            #for i in range(2, N+1):
            #    res = min(res, max(dp(K, N-i), dp(K-1, i-1)) + 1)
            #memo[(K, N)] = res

            # 用二分搜索代替线性搜索
            lo, hi = 1, N
            while lo <= hi:
                mid = (lo + hi) // 2
                broken = dp(K - 1, mid - 1)
                not_broken = dp(K, N - mid)
                if broken > not_broken:
                    hi = mid - 1
                    res = min(res, broken + 1)
                else:
                    lo = mid + 1
                    res = min(res, not_broken + 1)
            memo[(K, N)] = res
            return res

        return dp(K, N)